### ISSUE TITLE
Compute the MD5 hash of the file for Azure Storage

### ISSUE THREAD

**Issue by penguoir** [tus/tusd#1187]
**Is your feature request related to a problem? Please describe.**
I'm trying to use Ruby on Rails + Tusd. Rails' Active Storage requires a checksum to verify file integrity. Right now, Tusd doesn't compute the hash, so I have to disable the checksum verification on Rails, which is a cumbersome process.

**Describe the solution you'd like**
When saving blobs to Azure, save their MD5 hash too.

**Describe alternatives you've considered**

- Can generate the hash on Rails after uploading the file to Azure
- Can disable integrity verification

**Can you provide help with implementing this feature?**
Yes happy to help!

**Additional context**

This was mentioned when Azure storage was added:

https://github.com/tus/tusd/issues/401#issuecomment-913266909

Copied from that thread:

> may I ask what benefit computing the MD5 hash of the file has? (I've never used it, so I am curious) Would you compute the MD5 hash of the block of the blob, or the entire file?

> The hash is used to verify the integrity of the blob/file during transport.

Also, you can verify you don't have duplicates on your system.

**Comment by penguoir** [comment]
As a side note, I'd also like to save the metadata (currently saved under a separate blob) in the Azure-provided "metadata" field. But I can open a separate issue for that.

**Comment by penguoir** [comment]
Just had a look at implementing this:

- I don't think it's right to add the hash to the info blob. We write the info blob before reading the actual file, so there's no way to get the file's hash into the info blob (unless we update the info blob after the file is uploaded).
- I'm not sure whether it's possible to correctly compute the hash of the uploaded file. Depends on:
  - Does `UploadChunk` always run in start-to-finish order? Does it run in parallel or sequentially? We need it to run in start-to-finish order and sequentially for the hash function to work.
- How do we handle resumable uploads?

**Comment by Acconut** [comment]
Thanks for bringing this up. tusd currently doesn't have any feature for calculating or comparing checksums of the uploaded data, but I would like to change this in the future while adding support for the [tus checksum extension](https://tus.io/protocols/resumable-upload#checksum) and [HTTP digest fields](https://www.rfc-editor.org/rfc/rfc9530.html) (for [draft-ietf-httpbis-resumable-upload](https://www.ietf.org/archive/id/draft-ietf-httpbis-resumable-upload-04.html#name-integrity-digests)). I love to collaborate with you on this if you are interested.

Support for checksums / file integrity checks shouldn't be tied to Azure or any other storage. Instead, calculating the checksum while the data is uploaded is the responsibility of the central upload handling logic (in unrouted_handler.go). The calculated digests can then be used to verify the integrity of the entire upload or individual PATCH requests and be provided to the storage or hooks.

> * How do we handle resumable uploads?

We would have to store the state of the checksum calculation if the upload is interrupted/saved. If the upload is resumed, we can continue the calculation until the upload is finished.

**Comment by penguoir** [comment]
Sure I can spend a couple hours looking into this over the weekend and see how far I get.

**Comment by Acconut** [comment]
That sounds great! Before heading into an implementation, we can also brainstorm different implementation designs to make sure we cover all requirements.

**Comment by penguoir** [comment]
I've moved to an alternative storage provider that doesn't need a hash, so don't have appetite to work on this ticket. Sorry!

**Comment by Acconut** [comment]
No worries, thank you for the heads up!

**Comment by cfarmer-fearless** [comment]
Hello, hoping to open this thread up again.  I am a maintainer for [this project](https://github.com/CDCgov/data-exchange-upload/tree/main/upload-server), which uses tusd as a Golang library.  This project has the capability of copying files to secondary storage locations once upload completes, and I would like to use content hashes to verify the integrity of the file once it has been copied.

Some high level questions I have are...
- How should the hash of the file be generated?  You mention it can be done as chunks come across, but how does that work? 
- Seems like Azure and AWS have the capability to generate checksums asynchronously.  Should we rely on that?
- What specific HTTP headers need to be manipulated?
- Can this be done in hooks or does it need to be done deeper within tusd?
- What if the client wants to provide their own hash?
